src/core/thread/mle.cpp

```diff
@@ -165,7 +165,7 @@ otError Mle::Enable(void)
 
     UpdateLinkLocalAddress();
     SuccessOrExit(error = mSocket.Open(&Mle::HandleUdpReceive, this));
-    SuccessOrExit(error = mSocket.Bind(kUdpPort));
+    SuccessOrExit(error = mSocket.Bind(kUdpPort, true));
```

Comment:
We seem to already filter based on MLE port number in `Bind()` implementation.
```
#if OPENTHREAD_CONFIG_PLATFORM_UDP_ENABLE
     else if (!IsMlePort(aSocket.mSockName.mPort))
     {
         error = otPlatUdpBind(&aSocket);
```
 So do we even need this?

-------

In general, I am not sure if it is a good idea to add `bool aThreadOnly` falg to all `Bind()` methods, mainly since it seems to be only applicable when `OPENTHREAD_CONFIG_PLATFORM_UDP_ENABLE` is enabled.

I think by default, from the perspective of OT core and users, there should only one netif and it is the thread netif and OT sockets are tied to that interface only.

Reply:
No, it's not necessary here. However, I think it still meaningful to mark this should be a service on Thread interface only. This change is mainly for allowing running multiple OpenThread instances on a single host. Both TMF and MLE should run on Thread interface only. But border agent should run on all interfaces. The backbone agent should run on the backbone link interface(s) only.

I think we might need revert https://github.com/openthread/openthread/pull/5316, so that an OpenThread service can be bound to any host interface.

Reply:
BBR features will needs to use sockets that bind to all interfaces or backbone interface only. So, I think it's useful to use flags to indicate which interface to bind to.

Reply:
Here is how I see this:
- An OpenThread instance itself provides a "thread" netif to the user.
- All the socket APIs on an OT instance like `Socket::Bind()` should be tied to the provided "thread" netif only.
- I think OT core (representing an OT instance) should not be in general aware of any other netif (that is provided on device by host or other  instances of OT) and I think OT core should not be involved in managing (e.g., adding addresses, etc) on any other netif.
- So having a flag `aThreadOnly` in `Socket::Bind()` to indicate to bind  to thread netif vs all netifs inherently assumes an OT instance is aware of (and managing/binding to) more than one netif.

We have `OT_CONFIG_PLATFORM_UDP_ENABLE` which allow platforms to instead provide and handle UDP related socket APIs.
- Only in this context (when this feature is enabled), we sort of know of two netifs (one internal to OT core and one indirectly provided by platform UDP APIs).
- So only in this context, the `aThreadOnly` flags have any meaning and even in this case there seems to be no change in behavior in OT core code (related to `aThreadOnly` flag) and the flags is only passed down to platform UDP APIs.

I understand the use-case and reason what this PR is aiming to do, but I'd like to suggest that we try to add in a way that it is limited/used to when `OT_CONFIG_PLATFORM_UDP_ENABLE` feature is enabled.
- For example, as I pointed we already have in code logic that does filter MLE message based on the port number
```
#if OPENTHREAD_CONFIG_PLATFORM_UDP_ENABLE
     else if (!IsMlePort(aSocket.mSockName.mPort))
     {
         error = otPlatUdpBind(&aSocket);
```
- Instead of adding `aThreadOnly` to different OT APIs, can we consider adding additional/similar checks like the one for MLE which are scoped under `#if OPENTHREAD_CONFIG_PLATFORM_UDP_ENABLE` and take care of the special cases.

Thoughts?

Reply:
The `aThreadOnly` is not added as an internal argument which is not exposed via OT APIs.

Here's the use cases, it seems the core still need to know existence of other network interfaces.
* The backbone agent need to know the index of the backbone link interface.
* The Thread border agent should be listening on all network interfaces(Wi-Fi, Ethernet and so on) so that external commissioners can reach the agent through all PHYs.

Actually, in core/net/ip6, whether a datagram comes from the Thread network interface is also checked([code](https://github.com/openthread/openthread/blob/master/src/core/net/ip6.cpp#L1161)).

Besides, one motivation of platform UDP was to allow developers develop their apps/services directly on OpenThread if they find OpenThread's infrastructure is easier to write the code(For example, OpenThread has a good CoAP implementation) and the apps/services may be operating on other network interfaces just like the Thread border agent. Another motivation is trying to integrate OpenThread's IP stack with the host's, so that if OpenThread has a socket bound to 1000, the host should not bind to that port again and vice versa.

Thoughts?

Reply:
We do have the need to bind the the Backbone link so I think it is proper to have the argument for `Socket::Bind()` to choose which interface to bind to. 
But I also agree that "OT core (representing an OT instance) should not be in general aware of any other netif". 
That's why I suggested to use default argument `bool aThreadOnly = true` for the `Socket::Bind`.